### PR TITLE
[tc] Implement Predicates interface for bucket grouping suppo…

### PR DESCRIPTION
**Pull request by LeoLeo718**
[thin-client] Implement Predicates interface for bucket grouping support

## Problem Statement
The current Venice client lacks support for flexible data grouping and filtering capabilities, specifically:
- No built-in support for time-based data bucketing
- Limited ability to group numeric data into custom ranges
- Lack of unified interface for different types of predicates (int, long, time)

These limitations make it difficult for users to efficiently query and analyze data based on various grouping criteria.

## Solution
Implemented a comprehensive predicate system that supports:
1. Time bucket grouping:
   - Group data by time intervals (hourly, daily)
   - Support for custom time ranges
   - Efficient handling of timestamp-based queries

2. Custom bucket size grouping:
   - Flexible range-based grouping for long values
   - Support for arbitrary bucket boundaries
   - Optimized for common use cases like price ranges, metrics binning

3. Integer bucket grouping:
   - Dedicated support for integer-based grouping
   - Memory-efficient implementation compared to using longs
   - Suitable for scenarios like age groups, ratings, etc.

### Performance Considerations
- Used primitive types (int, long) to minimize boxing/unboxing
- Implemented efficient range checks
- Designed for minimal memory footprint

### Testing Done
- Comprehensive unit tests for all predicate types
- Integration tests with Venice storage system
- Edge case testing for boundary conditions
- Verified thread safety of implementations

### Code changes
- [x] Added new code behind **a config**: No new configs needed
- [x] Introduced new **log lines**: No new log lines added
  - [x] Confirmed if logs need to be **rate limited**: N/A

### **Concurrency-Specific Checks**
- [x] Code has **no race conditions** - all predicates are immutable
- [x] Proper **synchronization mechanisms** - not needed due to immutable design
- [x] No **blocking calls** - all operations are CPU-bound
- [x] Verified **thread-safe collections** - using immutable objects
- [x] Validated proper exception handling - all edge cases covered

## How was this PR tested?
- [x] New unit tests added:
  - `testTimeBucketGroupingPredicates`
  - `testCustomBucketSizeGrouping`
  - `testIntBucketGroupingPredicates`
- [x] New integration tests added:
  - Integration with Venice storage system
  - End-to-end data grouping tests
- [x] Modified existing tests to use new predicate interface
- [x] Verified backward compatibility with existing query patterns

## Does this PR introduce any user-facing or breaking changes?
- [x] Yes. 
  - Previous: Used `PredicateBuilder` for creating predicates
  - New: Direct use of `Predicate` class static methods
  - Impact: Cleaner API, better type safety, more intuitive usage
  - Migration: Simple replacement of import statements and method calls

### CODE REVIEW

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import com.linkedin.venice.annotation.Experimental;
+
+
+/**
+ * A specialized predicate interface for integer values.
+ * This interface provides factory methods for creating common integer comparison predicates
+ * such as equals, greater than, less than, etc.
+ */
+public interface IntPredicate extends Predicate<Integer> {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: Here we just have 2 type of numerical Predicate Int and Long, and I think we probably will also need Double, or might be even String.  So essentially we rely on `Comparable` interface for comparison and Object.equals() to achieve equalTo, greaterThan, greaterOrEqualls, lowerThan, lowerOrEquals, anyOf API. Wondering if we could reply on GreaterOrEqualsPredicate<T> to avoid duplicate codes, and we could convert any primitive type like int, long, double to Integer, Long, Double. This might also make user implemented compareTo and equal() to apply customerized filtering.  cc: @FelixGV

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/ComputeAggregationRequestBuilder.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.linkedin.venice.client.store;
+
+import com.linkedin.venice.client.store.predicate.Predicate;
+import java.util.Map;
+
+
+public interface ComputeAggregationRequestBuilder<K> extends ExecutableRequestBuilder<K, ComputeAggregationResponse> {
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/ComputeAggregationRequestBuilder.java]: This `ComputeAggregationRequestBuilder` might not be very meaningful here, as it is just interface, we could add its implementation and its test after.

**Review by haoxu07 — COMMENTED** [review]
Thanks for preparing this and the test. I think we might need to discuss how to define `Predicate` to see if we need to make it more general.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java (hunk 1 above)
Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: Thanks! I agree. I was thinking we could use a base ComparablePredicate interface to support more types.

Comment (FelixGV) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: Yes of course it is possible to use objects rather than primitives... The goal of using a primitive was to avoid boxing, since this is going to be a performance-sensitive code path.

I don't understand this part:

> This might also make user implemented compareTo and equal() to apply customerized filtering.

Can you elaborate or rephrase?

**Comment by FelixGV** [comment]
BTW, since we are getting started on this work, should I merge https://github.com/linkedin/venice/pull/1612 so that this PR can rebase cleanly on those changes rather than import (some of) them into this one?

**Comment by haoxu07** [comment]
> BTW, since we are getting started on this work, should I merge #1612 so that this PR can rebase cleanly on those changes rather than import (some of) them into this one?

Yeah, I think it is reasonable to merge this proposal.

**Comment by FelixGV** [comment]
> > BTW, since we are getting started on this work, should I merge https://github.com/linkedin/venice/pull/1612 so that this PR can rebase cleanly on those changes rather than import (some of) them into this one?
>
> Yeah, I think it is reasonable to merge this proposal.

Ok, looks like there is a merge conflict... let me rebase it, and then you can ship it...

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java (hunk 1 above)
Comment (haoxu07) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: As the Predicate we are going to implement are mainly based on comparison or equal check: 
GreaterOrEqualsPredicate
LessOrEqualsPredicate
AnyOf
...,

Currently, we might just need for Integer, Long, Double..., is it possible in the future: user define a class T, user implement the  `equals` and `compareTo` , then user expects comparison and equality check based Predicate for class T? (One example: T is a vector of Double )

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: > As the Predicate we are going to implement are mainly based on comparison or equal check: GreaterOrEqualsPredicate LessOrEqualsPredicate AnyOf ...,
> 
> Currently, we might just need for Integer, Long, Double..., is it possible in the future: user define a class T, user implement the `equals` and `compareTo` , then user expects comparison and equality check based Predicate for class T? (One example: T is a vector of Double )

I'm thinking about what if we maintain IntPredicate and LongPredicate for performance-critical scenarios while introducing a generic implementation based on the Comparable interface to support Double and custom user-defined types?

File: clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/predicate/PredicateTest.java [hunk 3]
```diff
@@ -346,4 +346,174 @@ public void testLongPredicates() {
...
+  public void testTimeBucketGroupingPredicates() {
+    Schema timeBucketSchema = new Schema.Parser().parse(
+        "{\n" + "  \"namespace\": \"example.predicate\",\n" + "  \"type\": \"record\",\n"
+            + "  \"name\": \"TimeBucketRecord\",\n" + "  \"fields\": [\n"
+            + "    { \"name\": \"timestamp\", \"type\": \"long\" },\n"
+            + "    { \"name\": \"value\", \"type\": \"string\" }\n" + "  ]\n" + "}");
+
+    // Create test records with fixed timestamps for better predictability
+    long baseTime = 1617235200000L; // 2021-04-01 00:00:00 UTC
+    long hourInMillis = 3600000L; // 1 hour in milliseconds
```
Comment (haoxu07) [inline clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/predicate/PredicateTest.java]: Nit: we could use TimeUnit.HOURS.toMillis(1) to make it more readable.

Comment (LeoLeo718) [reply clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/predicate/PredicateTest.java]: Yes, you are right, I already replaced that!

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java (hunk 1 above)
Comment (FelixGV) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: Ok I see… let’s break it up into a few separate aspects:

1. Should there be more predicates, including for floating points, floating point vectors, and even generic types? Yes, absolutely.
2. Should we allow users to define their own predicates? That one is more fraught… the VIP-5 design assumes all predicates are provided by the system, not by users.
3. But could custom predicates be done anyway? In theory, yes… but in practice, we do not want to run the users’ arbitrary code on the server-side. We could still allow custom predicates on the client-side only, but then it might have weird implications (e.g. The user adds one innocuous predicate, and their performance degrades because they get switched from server-side predicate push down to client-side evaluation).

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/IntPredicate.java]: > Ok I see… let’s break it up into a few separate aspects:
> 
> 1. Should there be more predicates, including for floating points, floating point vectors, and even generic types? Yes, absolutely.
> 2. Should we allow users to define their own predicates? That one is more fraught… the VIP-5 design assumes all predicates are provided by the system, not by users.
> 3. But could custom predicates be done anyway? In theory, yes… but in practice, we do not want to run the users’ arbitrary code on the server-side. We could still allow custom predicates on the client-side only, but then it might have weird implications (e.g. The user adds one innocuous predicate, and their performance degrades because they get switched from server-side predicate push down to client-side evaluation).

Thanks @FelixGV  for the detailed feedback. I agree that we should:
1.Keep the current implementation focused on system-provided predicates as per VIP-5 design
2.Plan for supporting more predicate types (floating points, vectors) in the future
3.Avoid user-defined predicates to maintain security and performance optimizations
I'll continue with the current approach of using primitive types for performance-critical scenarios. We can discuss the additional predicate types implementation.

File: clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/DoubleGreaterThanPredicate.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+public class DoubleGreaterThanPredicate implements DoublePredicate {
+  private final double threshold;
+
+  DoubleGreaterThanPredicate(double threshold) {
+    this.threshold = threshold;
+  }
+
+  @Override
+  public boolean evaluate(double value) {
+    return value > threshold;
```
Comment (haoxu07) [inline clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/DoubleGreaterThanPredicate.java]: There might be some issue here, as when run this check, we could encounter this     
```
    double a = 0.1d + 0.2d; // 0.30000000000000004
    double b = 0.3d;
    //a > b: true 
    // a - b: 5.551115123125783E-17
```


When we check with epsilon, we may encounter 

```
DoubleGreaterThanPredicate doubleGreaterThanPredicate = new DoubleEqualsPredicate(a);
doubleGreaterThanPredicate.evaluate(b); // True

DoubleEqualsPredicate doubleEqualsPredicate = new DoubleEqualsPredicate(a, 1e-15);
doubleEqualsPredicate.evaluate(b) // True
```

Should we add epsilon to greater or lower check to ensure mutual-exclusive?

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/DoubleGreaterThanPredicate.java]: That's pretty detailed, I think we can add an epsilon threshold for floating-point comparisons.
Just like:
public boolean evaluate(double value) {
        return (value - threshold) > epsilon;
        }
What do you think?

Comment (LeoLeo718) [reply clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/predicate/DoubleGreaterThanPredicate.java]: I modified all the comparison logic of float and double predicates
